mcs/class/Mono.Posix/Test/Mono.Unix.Native/SocketTest.cs

```diff
@@ -206,6 +206,651 @@ public void Shutdown ()
+				var actualAddress = new SockaddrIn ();
+				if (Syscall.getsockname (so1, actualAddress) < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				Assert.AreEqual (actualAddress.sa_family, UnixAddressFamily.AF_INET);
+				var port = Syscall.ntohs (actualAddress.sin_port);
+				Assert.IsTrue (port != 0);
+
+
+				var remoteAddress = new SockaddrIn {
+					sin_family = UnixAddressFamily.AF_INET,
+					sin_port = Syscall.htons (port),
+					sin_addr = new InAddr (127, 0, 0, 1),
+				};
+
+				// Send and receive a few bytes
+				long ret;
+				var buffer1 = new byte[] { 42, 43, 44 };
+				fixed (byte* ptr_buffer1 = buffer1) {
+					var iovecs1 = new Iovec[] {
+						new Iovec {
+							iov_base = (IntPtr) ptr_buffer1,
+							iov_len = (ulong) buffer1.Length,
+						},
+					};
+					var msghdr1 = new Msghdr {
+						msg_name = remoteAddress,
+						msg_iov = iovecs1,
+						msg_iovlen = 1,
+					};
+					ret = Syscall.sendmsg (so2, msghdr1, 0);
+					msghdr1.msg_name = remoteAddress.ToSockaddrStorage ();
+					if (ret >= 0)
+						ret = Syscall.sendmsg (so2, msghdr1, 0);
+				}
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+
+				var senderAddress = new SockaddrIn ();
+				var senderAddressStorage = new SockaddrStorage ();
+				var buffer2 = new byte[1024];
+				var buffer3 = new byte[1024];
+				fixed (byte* ptr_buffer2 = buffer2, ptr_buffer3 = buffer3) {
+					var iovecs2 = new Iovec[] {
+						new Iovec {
+							iov_base = (IntPtr) ptr_buffer2,
+							iov_len = (ulong) buffer2.Length,
+						},
+					};
+					var msghdr2 = new Msghdr {
+						msg_name = senderAddress,
+						msg_iov = iovecs2,
+						msg_iovlen = 1,
+					};
+					ret = Syscall.recvmsg (so1, msghdr2, 0);
+					msghdr2.msg_name = senderAddressStorage;
+					iovecs2[0].iov_base = (IntPtr) ptr_buffer3;
+					if (ret >= 0)
+						ret = Syscall.recvmsg (so1, msghdr2, 0);
+				}
+				if (ret < 0)
+					UnixMarshal.ThrowExceptionForLastError ();
+				Assert.AreEqual (senderAddress.sa_family, UnixAddressFamily.AF_INET);
+				Assert.AreEqual (senderAddress.sin_addr, new InAddr (127, 0, 0, 1));
+				var senderAddress2 = SockaddrIn.FromSockaddrStorage (senderAddressStorage);
+				Assert.AreEqual (senderAddress2.sa_family, UnixAddressFamily.AF_INET);
+				Assert.AreEqual (senderAddress2.sin_addr, new InAddr (127, 0, 0, 1));
+
+				Assert.AreEqual (buffer1.Length, ret);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer2[i]);
+				for (int i = 0; i < buffer1.Length; i++)
+					Assert.AreEqual (buffer1[i], buffer3[i]);
+			});
+		}
+
+		[Test]
+		public unsafe void ControlMsg ()
+		{
+			// Create two socket pairs and send inner_so1 and inner_so2 over the other socket pair using SCM_RIGHTS
+			WithSocketPair ((inner_so1, inner_so2) => {
+				WithSocketPair ((so1, so2) => {
+					// Create two SCM_RIGHTS control messages
+					var cmsg = new byte[2 * Syscall.CMSG_SPACE (sizeof (int))];
```

Comment:
Using the example in the [**cmsg**(3) man page](http://man7.org/linux/man-pages/man3/cmsg.3.html), I'm not entirely sure that this is correct. (It's *probably* correct, but it might be implementation specific.)

           struct msghdr msg = {0};
           struct cmsghdr *cmsg;
           int myfds[NUM_FD]; /* Contains the file descriptors to pass. */
           union {
               /* ancillary data buffer, wrapped in a union in order to ensure
                  it is suitably aligned */
               char buf[CMSG_SPACE(sizeof myfds)];
               struct cmsghdr align;
           } u;
           int *fdptr;

           msg.msg_control = u.buf;

`msghdr1.msg_control` (several lines below) is set to `cmsg`, meaning `cmsg` is equivalent to the man page's `u.buf` variable. Note that `u.buf` is a *union* between `char buf[CMSG_SPACE(sizeof myfds)]` *and* `struct msghdr`, meaning sizeof `u` is *at least* `sizeof(struct msghdr)` bytes in size, which would be `Cmsghdr.Size` bytes.

Not knowing what the value of `Cmsghdr.Size`, it is thus *plausible* (to idiots like me!) that `Cmsghdr.Size` is *greater than* what you actually allocated, `new byte[2*Syscall.CMSG_SPACE(4)]`.

This doesn't leave me with a good feeling.

I have to wonder if following the POSIX API too closely here is actually a good thing. This seems *brittle as hell*, *easy to get wrong*, not at all obvious, and I want to run screaming in terror. :-(